kasa/protocolfactory.py

```diff
@@ -0,0 +1,39 @@
+"""Module for protocol factory class."""
+from typing import Optional, Tuple, Type
+
+from .aestransport import AesTransport
+from .connectionparams import ConnectionParameters
+from .iotprotocol import IotProtocol
+from .klaptransport import KlapTransport, KlapTransportV2
+from .protocol import (
+    BaseTransport,
+    TPLinkProtocol,
+    TPLinkSmartHomeProtocol,
+    _XorTransport,
+)
+from .smartprotocol import SmartProtocol
+
+
+def get_protocol(
+    cparams: ConnectionParameters,
+) -> Optional[TPLinkProtocol]:
```

Comment:
As this is just using some connection parameters as input, it probably makes sense to have this as a method inside the class?

Reply:
Same as above for `device_factory` the reason for the `protocolfactory` is to avoid circular dependencies. Because the protocol construction needs to reference protocol classes but the protocol class is referencing the connection parameters in it's constructor so it can't go in the connection parameters class.  Unless you mean another class and I'm missing something.

Reply:
I'm currently rather busy to look it up, but couldn't we make the imports inside the function to avoid the circular deps?